gravitational/teleport
# Ensure Connect My Computer agent is terminated if orphaned

This draft PR includes a WIP implementation of the agent cleanup daemon. This is based on the implementation discussed in #29396.

## How the agent cleanup daemon works

The first bunch of commits adds the implementation of the agent cleanup daemon itself. The way it works is that the main process of the Electron app starts both a Teleport agent (for Connect My Computer) and an agent cleanup daemon, as separate processes. The sole purpose of the daemon is to stop the agent if the parent, the main Electron process, gets terminated.

How do we know that the parent got terminated? When forking a child process from a parent process, Node.js establishes an IPC channel between them for communication. In the agent cleanup daemon, we can observe the channel being closed/dirupted by listening for [the `disconnect` event on `process`](https://nodejs.org/docs/latest-v18.x/api/process.html#event-disconnect). Since we never purposefully close that channel, when the cleanup daemon detects that it was closed, it can assume that the parent has died. This is similar to [how you'd usually approach this problem on Unix](https://stackoverflow.com/a/42924532/742872).

There are a couple of nuances, as explained in `agentCleanupDaemon.js`, but in a gist this is how it works.

## How the agent cleanup daemon integrates with Connect My Computer

The next bunch of commits makes it so that whenever Connect spawns a Teleport agent, it also spawns an accompanying cleanup daemon. The rest of the app doesn't even need to know about it, so we can 100% depend on child process events there.

## How to verify that it works

First of, you need to be able to launch the setup of Connect My Computer. Check out #29727 to see how to do this.

Next, launch master version of the app, start the setup and open Activity Monitor. View -> My Processes, search for "teleport". Double-click it. It should say what the parent process is and what its PID is. Send SIGKILL to the parent process (`kill -s KILL <parent pid>`).

You should see that the teleport process is still running, despite the fact that Connect is now closed. This is bad, because the agent is still making your computer available in the cluster.

Switch to this branch and repeat the process. You should see that the agent gets killed soon after you kill the main process. If you started Connect with `yarn start-term`, you should even see logs from the cleanup daemon in the terminal.

TODO:

   - I have to plug Winston _inside_ the cleanup daemon. At the moment, all of our logging is routed through the main process of the Electron app. But the cleanup daemon is made specifically for situations where the main process dies, so it doesn't make sense to route daemon's logs through the main process.
